### ISSUE TITLE
validate does not run faster when using the everyN option

### ISSUE THREAD

**Issue by Yeqzids** [NASA-PDS-Incubator/validate#661]
### Checked for duplicates

Yes - I've already checked

### 🐛 Describe the bug

Hi all. I am running this on a large dataset (~1700 files) and there is something that I don't understand. I'd expect that --everyN will run faster since it only checks the content for the Nth file (#1). However, for my case the test with --everyN and the one that without takes roughly the same amount of time to run (in my case, about 3 hours, with a difference of about 10 seconds). Is this expected? I am using 3.2.0 and a large everyN (100).

### 🕵️ Expected behavior

I expected validate to run significantly faster when using a large everyN.

### 📜 To Reproduce

Here is the command that I use:

`validate-3.2.0/bin/validate bundle_test_v0.0.xml -S PDS4_EBT_*.sch -x PDS4_*.xsd --skip-context-validation -R pds4.bundle --everyN 100`

### 🖥 Environment Info

- Version of this software: 3.2.0
- Operating System: Red Hat 8.7

**Comment by kbowley-asu** [comment]
I am very interested in this issue as we can spot check individual products, but have to use --skip-content-validation when checking a collection due to the fact that even our small collections have over 15,000 files.

**Comment by jordanpadams** [comment]
@Yeqzids we will take a look. In the meantime, how big are the contents of the files you are trying to validate? Most of the performance degradation is actually in the schema/schematron validation, for small products, content validation is completed in seconds (sometimes less). We have some metrics here: https://nasa-pds.github.io/validate/operate/index.html#Metrics

also, depending on your operations scenario, there are numerous other ways you can improve performance, while also executing validation in full on the products, like running validate in parallel.

https://nasa-pds.github.io/validate/operate/index.html#Advanced_Bundle_Validation

Given `my_bundle.xml`
```
# Execute bundle validation but only check bundle/collection referential integrity and ignore the data products
validate --skip-product-validation --rule pds4.bundle --target my_bundle.xml

# Point at several targets with different validate runs for parallelizing content validation
validate --rule pds4.label --target my_bundle/document/*.xml
validate --rule pds4.label --target my_bundle/release1/*.xml
validate --rule pds4.label --target my_bundle/release2/*.xml
validate --rule pds4.label --target my_bundle/release3/*.xml

# Or you could do the same thing with target manifests
validate --rule pds4.label --target-manifest documents.txt
validate --rule pds4.label --target-manifest release1_data.txt
validate --rule pds4.label --target-manifest release2_data.txt
validate --rule pds4.label --target-manifest release3_data.txt
```

If you are running on a linux system you can also execute `validate-bundle` using GNU Parallels: https://nasa-pds.github.io/validate/operate/index.html#Using_validate-bundle_parallelization

The biggest thing here is, in the end, we still want to validate every bit at some point before it is archived. Happy to tag up and discuss your operations pipeline architecture as well if there are ways we can help improve validate's performance.

**Comment by Yeqzids** [comment]
Jordan -- appreciate the quick response. Thanks for the tips on parallelization -- they are very helpful and I will look into them.

> In the meantime, how big are the contents of the files you are trying to validate?

About half of them are FITS data that are 200-450 MB each. Do I understand it correctly that with --everyN, the content of these (N-1) files will be skipped? Looking at the processes, the CPUs are at 100% most/all the time so it seems the program is doing some computing-heavy work, rather than reading and/or passing over files.

**Comment by al-niessner** [comment]
@jordanpadams @Yeqzids 

Are the files being skipped with everyN large or small files? If large where content check is largest part of validation, then you should see a speedup. If small files where content check time is seconds, then would expect little change with everyN because only content validation is skipped allowing other slow parts of validate to dominate that process all instead of everyN.

When I did the initial implementation of everyN, tested with KB size tables and saw no improvement. Changed to GB sized files and noticed a significant speed up.

Best way to determine if any speedup from everyN is to be expected is to validate the product (not bundle or collection) then validate it again with `--skip-content-validation` and look at the time difference.

**Comment by Yeqzids** [comment]
Hi @al-niessner -- sorry for the late response. 50% of the files in my bundle are big FITS files, so an improvement should be seen as long as --everyN >> 2.

Interesting, it seems I can't get @jordanpadams 's suggestion of using --target to work. When I run `validate-3.2.0/bin/validate bundle_test_v0.0.xml -S PDS4_EBT_*.sch -x PDS4_*.xsd --skip-context-validation -R pds4.bundle --target folder1/*.xml`, the "Targets" output picks up the right set of files, but the program still proceeds to check files in other folders.

**Comment by al-niessner** [comment]
The -R pds4.bundle forces a directory scan. Change to pds4.label or
something like that.

It could be pathological in that you are getting only the files that use
all of the time. Doubt it; especially because your N was big if i remember
correctly.

**Comment by jordanpadams** [comment]
> validate-3.2.0/bin/validate bundle_test_v0.0.xml -S PDS4_EBT_*.sch -x PDS4_*.xsd --skip-context-validation -R pds4.bundle --target folder1/*.xml

@Yeqzids for this use case, try:

```
validate-3.2.0/bin/validate bundle_test_v0.0.xml -S PDS4_EBT_*.sch -x PDS4_*.xsd --skip-context-validation -R pds4.label --target folder1/*.xml
```
note the use of `pds4.label` rule

**Comment by Yeqzids** [comment]
Ahh, thank you @al-niessner @jordanpadams . Now I can get both `--target` and `--everyN` to work! :):) But my original intent was to use `pds4.bundle` while doing `--everyN`. Sounds like `-R pds4.bundle` would just overrides `--everyN` or `--target` without a warning (or I missed it somewhere). Would it be possible to force `--everyN` when using `-R pds4.bundle`?

**Comment by Yeqzids** [comment]
Ben Hirsch suggests that one can run two commands -- one with `-R pds4.bundle` and `--skip-content-validation` for an overall integration check, and the other with `-R pds4.label` and `--everyN` (or `--target`) to check the every Nth file in the bundle. My issue is solved, and I am closing the ticket. Thank you all!

TL; DR: `--everyN` only works with `-R pds4.label`.

**Comment by jordanpadams** [comment]
@Yeqzids correct. I think the best possible solution is to break up the content validation, and run the referential integrity checks (`-R pds4.bundle`) separately with content validation turned off.
